<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>TidyMultiqc • TidyMultiqc</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="TidyMultiqc">
<meta property="og:description" content="TidyMultiqc">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">TidyMultiqc</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.0.3</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../articles/TidyMultiqc.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/multimeric/TidyMultiqc/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="TidyMultiqc_files/accessible-code-block-0.0.1/empty-anchor.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>TidyMultiqc</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/multimeric/TidyMultiqc/blob/HEAD/vignettes/TidyMultiqc.Rmd" class="external-link"><code>vignettes/TidyMultiqc.Rmd</code></a></small>
      <div class="hidden name"><code>TidyMultiqc.Rmd</code></div>

    </div>

    
    
<div class="section level2">
<h2 id="test-data">Test Data<a class="anchor" aria-label="anchor" href="#test-data"></a>
</h2>
<p>In this example we’ll be using the WGS example report used in the MultiQC documentation. You can find view the report in your browser <a href="https://multiqc.info/examples/wgs/multiqc_report.html" class="external-link">here</a>, and find the associated <code>multiqc_data.json</code> file <a href="https://github.com/ewels/MultiQC_website/blob/master/public_html/examples/wgs/multiqc_data/multiqc_data.json" class="external-link">here</a>. Feel free to download this file and use it to follow along with this vignette.</p>
<p>In the rest of the vignette we will use the <code>multiqc_data_path</code> variable to indicate the path to this <code>multiqc_data.json</code> file. Feel free to set this variable to the path to this file on your system.</p>
</div>
<div class="section level2">
<h2 id="setup">Setup<a class="anchor" aria-label="anchor" href="#setup"></a>
</h2>
<p>First, install the package using:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"TidyMultiqc"</span><span class="op">)</span></span></code></pre></div>
<p>Then if you want, you can load the package:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://multimeric.github.io/TidyMultiqc/" class="external-link">TidyMultiqc</a></span><span class="op">)</span></span></code></pre></div>
<p>However, for the sake of this tutorial we will explicitly use namespaced functions, so we won’t be using <code>library</code> a great deal.</p>
</div>
<div class="section level2">
<h2 id="basic-usage">Basic Usage<a class="anchor" aria-label="anchor" href="#basic-usage"></a>
</h2>
<p>The main entry point to the <code>TidyMultiqc</code> package is the function <code>load_multiqc</code>. A basic invocation of this function looks like this:</p>
<p>(note the arrow for scrolling through columns)</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">df</span> <span class="op">=</span> <span class="fu">TidyMultiqc</span><span class="fu">::</span><span class="fu"><a href="../reference/load_multiqc.html">load_multiqc</a></span><span class="op">(</span><span class="va">multiqc_data_path</span><span class="op">)</span></span>
<span><span class="va">df</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 6 × 165</span></span></span>
<span><span class="co">#&gt;   metadata.sam…¹ gener…² gener…³ gener…⁴ gener…⁵ gener…⁶ gener…⁷ gener…⁸ gener…⁹</span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>            <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> P4107_1003      8.68<span style="color: #949494;">e</span>8  8.48<span style="color: #949494;">e</span>8    97.6      40     365    41.4    92.3    92.2</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span> P4107_1004      1.00<span style="color: #949494;">e</span>9  9.85<span style="color: #949494;">e</span>8    98.2      46     363    41.0    92.3    92.2</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">3</span> P4107_1005      9.75<span style="color: #949494;">e</span>8  9.56<span style="color: #949494;">e</span>8    98.0      45     368    41.2    92.3    92.2</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">4</span> P4107_1002      8.66<span style="color: #949494;">e</span>8  8.47<span style="color: #949494;">e</span>8    97.8      40     367    41.3    92.3    92.2</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">5</span> P4107_1006      9.12<span style="color: #949494;">e</span>8  8.95<span style="color: #949494;">e</span>8    98.1      43     362    41.3    92.3    92.2</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">6</span> P4107_1001      7.72<span style="color: #949494;">e</span>8  7.51<span style="color: #949494;">e</span>8    97.3      36     358    41.4    92.3    92.2</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># … with 156 more variables: general.10_x_pc &lt;dbl&gt;, general.30_x_pc &lt;dbl&gt;,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   general.50_x_pc &lt;dbl&gt;, general.genome &lt;chr&gt;,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   general.number_of_variants_before_filter &lt;dbl&gt;,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   general.number_of_known_variants_brie_non_empty_id &lt;dbl&gt;,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   general.number_of_known_variants_brie_non_empty_id_percent &lt;dbl&gt;,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   general.number_of_effects &lt;dbl&gt;, general.genome_total_length &lt;dbl&gt;,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   general.genome_effective_length &lt;dbl&gt;, general.change_rate &lt;dbl&gt;, …</span></span></span></code></pre></div>
<p>We’ve now generated a <code>tibble</code> (a kind of data frame), whose rows are samples in the QC report, and whose columns are QC data and metadata about these samples.</p>
<p>By default this function only returns the “general” statistics, which are the ones in the “General Statistics” table at the top of the MultiQC report. In TidyMultiqc, these statistics are all prefixed by <code>general.</code> We can also extract the “raw” statistics, which includes some fields normally hidden from the report. These statistics will have the prefix <code>raw.&lt;toolname&gt;.</code> where <code>&lt;toolname&gt;</code> is the QC tool used to calculate it.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">TidyMultiqc</span><span class="fu">::</span><span class="fu"><a href="../reference/load_multiqc.html">load_multiqc</a></span><span class="op">(</span><span class="va">multiqc_data_path</span>, sections <span class="op">=</span> <span class="st">'raw'</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 6 × 224</span></span></span>
<span><span class="co">#&gt;   metadata.sam…¹ raw.q…² raw.q…³ raw.q…⁴ raw.q…⁵ raw.q…⁶ raw.q…⁷ raw.q…⁸ raw.q…⁹</span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>          <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> P4107_1003     /scrat…  8.68<span style="color: #949494;">e</span>8  8.48<span style="color: #949494;">e</span>8 1.25<span style="color: #949494;">e</span>11 1.25<span style="color: #949494;">e</span>11  <span style="text-decoration: underline;">28</span>161.     363    50.3</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span> P4107_1004     /scrat…  1.00<span style="color: #949494;">e</span>9  9.85<span style="color: #949494;">e</span>8 1.46<span style="color: #949494;">e</span>11 1.46<span style="color: #949494;">e</span>11  <span style="text-decoration: underline;">28</span>253.     361    50.4</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">3</span> P4107_1005     /scrat…  9.75<span style="color: #949494;">e</span>8  9.56<span style="color: #949494;">e</span>8 1.41<span style="color: #949494;">e</span>11 1.41<span style="color: #949494;">e</span>11  <span style="text-decoration: underline;">30</span>723.     365    50.3</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">4</span> P4107_1002     /scrat…  8.66<span style="color: #949494;">e</span>8  8.47<span style="color: #949494;">e</span>8 1.25<span style="color: #949494;">e</span>11 1.25<span style="color: #949494;">e</span>11  <span style="text-decoration: underline;">29</span>317.     364    50.3</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">5</span> P4107_1006     /scrat…  9.12<span style="color: #949494;">e</span>8  8.95<span style="color: #949494;">e</span>8 1.32<span style="color: #949494;">e</span>11 1.32<span style="color: #949494;">e</span>11  <span style="text-decoration: underline;">30</span>869.     360    50.3</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">6</span> P4107_1001     /scrat…  7.72<span style="color: #949494;">e</span>8  7.51<span style="color: #949494;">e</span>8 1.11<span style="color: #949494;">e</span>11 1.11<span style="color: #949494;">e</span>11  <span style="text-decoration: underline;">28</span>932.     355    50.3</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># … with 215 more variables:</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   raw.qualimap_bamqc_genome_results.mean_coverage &lt;dbl&gt;,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   raw.qualimap_bamqc_genome_results.percentage_aligned &lt;dbl&gt;,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   raw.snpeff.genome &lt;chr&gt;, raw.snpeff.number_of_variants_before_filter &lt;dbl&gt;,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   raw.snpeff.empty_id &lt;dbl&gt;, raw.snpeff.empty_id_percent &lt;dbl&gt;,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   raw.snpeff.number_of_effects &lt;dbl&gt;, raw.snpeff.genome_total_length &lt;dbl&gt;,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   raw.snpeff.genome_effective_length &lt;dbl&gt;, raw.snpeff.change_rate &lt;dbl&gt;, …</span></span></span></code></pre></div>
<p>Often you won’t care about fields like <code>raw.qualimap_bamqc_genome_results.bam_file</code>, the path to the original BAM file, but ‘raw’ at least provides this option.</p>
<p>You can also combine both <code>general</code> and <code>raw</code> sections by passing in a longer vector:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">df_both</span> <span class="op">=</span> <span class="fu">TidyMultiqc</span><span class="fu">::</span><span class="fu"><a href="../reference/load_multiqc.html">load_multiqc</a></span><span class="op">(</span><span class="va">multiqc_data_path</span>, sections <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'raw'</span>, <span class="st">'general'</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">df_both</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 388</span></span></code></pre></div>
<p>That’s a lot of columns!</p>
</div>
<div class="section level2">
<h2 id="uses">Uses<a class="anchor" aria-label="anchor" href="#uses"></a>
</h2>
<p>This section will briefly talk about some downstream use-cases for this package.</p>
<div class="section level3">
<h3 id="plotting">Plotting<a class="anchor" aria-label="anchor" href="#plotting"></a>
</h3>
<p>One use for this data frame is creating QC plots. For example, to visualise the duplication rate per sample:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://magrittr.tidyverse.org" class="external-link">magrittr</a></span><span class="op">)</span></span>
<span></span>
<span><span class="va">df</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">metadata.sample_id</span>, y<span class="op">=</span><span class="va">general.percent_duplication</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_bar.html" class="external-link">geom_col</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="TidyMultiqc_files/figure-html/unnamed-chunk-6-1.png" width="700"></p>
<p>Of course, this is basically just replicating a plot already in the MultiQC report, but now we can customise it how we like!</p>
</div>
<div class="section level3">
<h3 id="hypothesis-testing">Hypothesis Testing<a class="anchor" aria-label="anchor" href="#hypothesis-testing"></a>
</h3>
<p>With all this data, we might also want to test a specific hypothesis! For example, we might want to test the hypothesis that the mean GC content is the same as the mean GC content in the human genome (41%). If we assume that GC content is normally distributed, we can do the following:</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/stats/t.test.html" class="external-link">t.test</a></span><span class="op">(</span><span class="va">df</span><span class="op">$</span><span class="va">general.percent_gc</span>, mu<span class="op">=</span><span class="fl">41</span><span class="op">)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;  One Sample t-test</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; data:  df$general.percent_gc</span></span>
<span><span class="co">#&gt; t = -1, df = 5, p-value = 0.3632</span></span>
<span><span class="co">#&gt; alternative hypothesis: true mean is not equal to 41</span></span>
<span><span class="co">#&gt; 95 percent confidence interval:</span></span>
<span><span class="co">#&gt;  40.40490 41.26176</span></span>
<span><span class="co">#&gt; sample estimates:</span></span>
<span><span class="co">#&gt; mean of x </span></span>
<span><span class="co">#&gt;  40.83333</span></span></code></pre></div>
<p>It seems that we cannot reject this hypothesis, so these may well be human samples!</p>
</div>
</div>
<div class="section level2">
<h2 id="extracting-metadata">Extracting Metadata<a class="anchor" aria-label="anchor" href="#extracting-metadata"></a>
</h2>
<p>It may be the case that your samples have important metadata that you want in your data frame. For example, it is common for the sample names or file names to be composed of a number of metadata fields, and indeed the report we are working with has IDs such as <code>P4107_1003</code>, which is composed of two identifiers.</p>
<p>To include this metadata in our output, we need to provide the <code>find_metadata</code> argument to <code>load_multiqc</code>, which is a function that is called for each sample, and which returns a named vector of metadata fields for that sample. It also gets passed the entire parsed MultiQC JSON report, so the function can traverse the structure as it wants to extract metadata.</p>
<div class="section level3">
<h3 id="from-the-file-name">From the File Name<a class="anchor" aria-label="anchor" href="#from-the-file-name"></a>
</h3>
<p>Here is an example that parses the input file names to annotate additional metadata. Notice that the first argument our function is passed is a string which is the sample identifier for a sample, and how in this example we ignore the <code>parsed</code> argument. Also notice that the names we give to our return value (“batch” and “sample”) are prefixed by “metadata” to become the final names in the data frame.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">TidyMultiqc</span><span class="fu">::</span><span class="fu"><a href="../reference/load_multiqc.html">load_multiqc</a></span><span class="op">(</span></span>
<span>  <span class="va">multiqc_data_path</span>, </span>
<span>  find_metadata <span class="op">=</span> <span class="kw">function</span><span class="op">(</span><span class="va">sample</span>, <span class="va">parsed</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="co"># Split the sample ID to obtain some metadata</span></span>
<span>    <span class="va">segments</span> <span class="op">&lt;-</span> <span class="fu">stringr</span><span class="fu">::</span><span class="fu"><a href="https://stringr.tidyverse.org/reference/str_split.html" class="external-link">str_split</a></span><span class="op">(</span><span class="va">sample</span>, <span class="st">"_"</span><span class="op">)</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>      batch <span class="op">=</span> <span class="va">segments</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>,</span>
<span>      sample <span class="op">=</span> <span class="va">segments</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span></span>
<span>    <span class="op">)</span></span>
<span>  <span class="op">}</span></span>
<span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 6 × 167</span></span></span>
<span><span class="co">#&gt;   metadata.sam…¹ metad…² metad…³ gener…⁴ gener…⁵ gener…⁶ gener…⁷ gener…⁸ gener…⁹</span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>          <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> P4107_1003     P4107   1003     8.68<span style="color: #949494;">e</span>8  8.48<span style="color: #949494;">e</span>8    97.6      40     365    41.4</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span> P4107_1004     P4107   1004     1.00<span style="color: #949494;">e</span>9  9.85<span style="color: #949494;">e</span>8    98.2      46     363    41.0</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">3</span> P4107_1005     P4107   1005     9.75<span style="color: #949494;">e</span>8  9.56<span style="color: #949494;">e</span>8    98.0      45     368    41.2</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">4</span> P4107_1002     P4107   1002     8.66<span style="color: #949494;">e</span>8  8.47<span style="color: #949494;">e</span>8    97.8      40     367    41.3</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">5</span> P4107_1006     P4107   1006     9.12<span style="color: #949494;">e</span>8  8.95<span style="color: #949494;">e</span>8    98.1      43     362    41.3</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">6</span> P4107_1001     P4107   1001     7.72<span style="color: #949494;">e</span>8  7.51<span style="color: #949494;">e</span>8    97.3      36     358    41.4</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># … with 158 more variables: general.1_x_pc &lt;dbl&gt;, general.5_x_pc &lt;dbl&gt;,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   general.10_x_pc &lt;dbl&gt;, general.30_x_pc &lt;dbl&gt;, general.50_x_pc &lt;dbl&gt;,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   general.genome &lt;chr&gt;, general.number_of_variants_before_filter &lt;dbl&gt;,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   general.number_of_known_variants_brie_non_empty_id &lt;dbl&gt;,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   general.number_of_known_variants_brie_non_empty_id_percent &lt;dbl&gt;,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   general.number_of_effects &lt;dbl&gt;, general.genome_total_length &lt;dbl&gt;,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   general.genome_effective_length &lt;dbl&gt;, general.change_rate &lt;dbl&gt;, …</span></span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="from-the-file-path">From the File Path<a class="anchor" aria-label="anchor" href="#from-the-file-path"></a>
</h3>
<p>We can extend this approach, but this time actually look up the file paths within the <code>report_data_sources</code> section of the MultiQC report:</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">TidyMultiqc</span><span class="fu">::</span><span class="fu"><a href="../reference/load_multiqc.html">load_multiqc</a></span><span class="op">(</span></span>
<span>  <span class="va">multiqc_data_path</span>, </span>
<span>  find_metadata <span class="op">=</span> <span class="kw">function</span><span class="op">(</span><span class="va">sample</span>, <span class="va">parsed</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="co"># This gives us the path to the fastqc output file</span></span>
<span>    <span class="va">filepath</span> <span class="op">=</span> <span class="va">parsed</span><span class="op">$</span><span class="va">report_data_sources</span><span class="op">$</span><span class="va">FastQC</span><span class="op">$</span><span class="va">all_sections</span><span class="op">[[</span><span class="va">sample</span><span class="op">]</span><span class="op">]</span></span>
<span>    <span class="co"># Split into path segments</span></span>
<span>    <span class="va">path_segments</span> <span class="op">=</span> <span class="fu">stringr</span><span class="fu">::</span><span class="fu"><a href="https://stringr.tidyverse.org/reference/str_split.html" class="external-link">str_split</a></span><span class="op">(</span><span class="va">filepath</span>, <span class="st">"/"</span><span class="op">)</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span></span>
<span>    <span class="co"># The filename is the last path segment</span></span>
<span>    <span class="va">filename</span> <span class="op">=</span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/nth.html" class="external-link">last</a></span><span class="op">(</span><span class="va">path_segments</span><span class="op">)</span></span>
<span>    <span class="co"># Split the filename using dots and underscores</span></span>
<span>    <span class="va">name_segments</span> <span class="op">=</span> <span class="fu">stringr</span><span class="fu">::</span><span class="fu"><a href="https://stringr.tidyverse.org/reference/str_split.html" class="external-link">str_split</a></span><span class="op">(</span><span class="va">filename</span>, <span class="st">"[_\\.]"</span><span class="op">)</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span></span>
<span>    <span class="co"># Arbitrarily assign names for the outputs</span></span>
<span>    <span class="va">name_segments</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu">purrr</span><span class="fu">::</span><span class="fu"><a href="https://rlang.r-lib.org/reference/set_names.html" class="external-link">set_names</a></span><span class="op">(</span><span class="va">LETTERS</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">name_segments</span><span class="op">)</span><span class="op">]</span><span class="op">)</span></span>
<span>  <span class="op">}</span></span>
<span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 6 × 173</span></span></span>
<span><span class="co">#&gt;   metadata.sam…¹ metad…² metad…³ metad…⁴ metad…⁵ metad…⁶ metad…⁷ metad…⁸ metad…⁹</span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>          <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>  </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> P4107_1003     P4107   1003    S3      L004    R1      001     fastqc  zip    </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span> P4107_1004     P4107   1004    S4      L005    R1      001     fastqc  zip    </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">3</span> P4107_1005     P4107   1005    S5      L006    R2      001     fastqc  zip    </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">4</span> P4107_1002     P4107   1002    S2      L003    R2      001     fastqc  zip    </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">5</span> P4107_1006     P4107   1006    S6      L007    R2      001     fastqc  zip    </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">6</span> P4107_1001     P4107   1001    S1      L002    R2      001     fastqc  zip    </span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># … with 164 more variables: general.total_reads &lt;dbl&gt;,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   general.mapped_reads &lt;dbl&gt;, general.percentage_aligned &lt;dbl&gt;,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   general.median_coverage &lt;int&gt;, general.median_insert_size &lt;int&gt;,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   general.avg_gc &lt;dbl&gt;, general.1_x_pc &lt;dbl&gt;, general.5_x_pc &lt;dbl&gt;,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   general.10_x_pc &lt;dbl&gt;, general.30_x_pc &lt;dbl&gt;, general.50_x_pc &lt;dbl&gt;,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   general.genome &lt;chr&gt;, general.number_of_variants_before_filter &lt;dbl&gt;,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   general.number_of_known_variants_brie_non_empty_id &lt;dbl&gt;, …</span></span></span></code></pre></div>
<p>Of course in a real application we would choose specific names for each field.</p>
</div>
<div class="section level3">
<h3 id="from-multiqc-configuration">From MultiQC Configuration<a class="anchor" aria-label="anchor" href="#from-multiqc-configuration"></a>
</h3>
<p>Finally, we might want to include metadata that doesn’t relate to the sample at all. For example, MultiQC has a number of report fields prefixed by <code>config_</code> that we might want to store:</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">TidyMultiqc</span><span class="fu">::</span><span class="fu"><a href="../reference/load_multiqc.html">load_multiqc</a></span><span class="op">(</span></span>
<span>  <span class="va">multiqc_data_path</span>, </span>
<span>  find_metadata <span class="op">=</span> <span class="kw">function</span><span class="op">(</span><span class="va">sample</span>, <span class="va">parsed</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="va">parsed</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>      <span class="st">"config_creation_date"</span>,</span>
<span>      <span class="st">"config_version"</span></span>
<span>    <span class="op">)</span><span class="op">]</span></span>
<span>  <span class="op">}</span></span>
<span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 6 × 167</span></span></span>
<span><span class="co">#&gt;   metadata.sam…¹ metad…² metad…³ gener…⁴ gener…⁵ gener…⁶ gener…⁷ gener…⁸ gener…⁹</span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>          <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> P4107_1003     2021-0… 1.10     8.68<span style="color: #949494;">e</span>8  8.48<span style="color: #949494;">e</span>8    97.6      40     365    41.4</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span> P4107_1004     2021-0… 1.10     1.00<span style="color: #949494;">e</span>9  9.85<span style="color: #949494;">e</span>8    98.2      46     363    41.0</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">3</span> P4107_1005     2021-0… 1.10     9.75<span style="color: #949494;">e</span>8  9.56<span style="color: #949494;">e</span>8    98.0      45     368    41.2</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">4</span> P4107_1002     2021-0… 1.10     8.66<span style="color: #949494;">e</span>8  8.47<span style="color: #949494;">e</span>8    97.8      40     367    41.3</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">5</span> P4107_1006     2021-0… 1.10     9.12<span style="color: #949494;">e</span>8  8.95<span style="color: #949494;">e</span>8    98.1      43     362    41.3</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">6</span> P4107_1001     2021-0… 1.10     7.72<span style="color: #949494;">e</span>8  7.51<span style="color: #949494;">e</span>8    97.3      36     358    41.4</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># … with 158 more variables: general.1_x_pc &lt;dbl&gt;, general.5_x_pc &lt;dbl&gt;,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   general.10_x_pc &lt;dbl&gt;, general.30_x_pc &lt;dbl&gt;, general.50_x_pc &lt;dbl&gt;,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   general.genome &lt;chr&gt;, general.number_of_variants_before_filter &lt;dbl&gt;,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   general.number_of_known_variants_brie_non_empty_id &lt;dbl&gt;,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   general.number_of_known_variants_brie_non_empty_id_percent &lt;dbl&gt;,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   general.number_of_effects &lt;dbl&gt;, general.genome_total_length &lt;dbl&gt;,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   general.genome_effective_length &lt;dbl&gt;, general.change_rate &lt;dbl&gt;, …</span></span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="extracting-plot-data">Extracting Plot Data<a class="anchor" aria-label="anchor" href="#extracting-plot-data"></a>
</h2>
<div class="section level3">
<h3 id="motivation">Motivation<a class="anchor" aria-label="anchor" href="#motivation"></a>
</h3>
<p>It is occasionally useful to extract QC data from the MultiQC plots. For example, let’s say we want to calculate the median quality score of every base in each sample. Unfortunately, MultiQC provides no numerical summary statistic for the read quality, it only has mapping quality and pass/fails for the per-base sequence quality:</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">df_both</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://tidyselect.r-lib.org/reference/starts_with.html" class="external-link">contains</a></span><span class="op">(</span><span class="st">'quality'</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 6 × 5</span></span></span>
<span><span class="co">#&gt;   raw.qualimap_bamqc_genome_results.mean_mappi…¹ raw.f…² raw.f…³ raw.f…⁴ raw.f…⁵</span></span>
<span><span class="co">#&gt;                                            <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>  </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span>                                           50.3       0 pass    warn    pass   </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span>                                           50.4       0 pass    fail    pass   </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">3</span>                                           50.3       0 pass    warn    pass   </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">4</span>                                           50.3       0 pass    fail    pass   </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">5</span>                                           50.3       0 pass    warn    pass   </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">6</span>                                           50.3       0 pass    warn    pass   </span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># … with abbreviated variable names</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   ¹​raw.qualimap_bamqc_genome_results.mean_mapping_quality,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   ²​raw.fastqc.sequences_flagged_as_poor_quality,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   ³​raw.fastqc.per_base_sequence_quality,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   ⁴​raw.fastqc.per_tile_sequence_quality,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   ⁵​raw.fastqc.per_sequence_quality_scores</span></span></span></code></pre></div>
<p>However, our MultiQC report does have <em>plots</em> that contain this information. In particular, let’s look at the “Per Sequence Quality Scores” plot.</p>
</div>
<div class="section level3">
<h3 id="listing-plots">Listing Plots<a class="anchor" aria-label="anchor" href="#listing-plots"></a>
</h3>
<p>Firstly, we need the ID of the plot we want. This isn’t necessarily obvious from just looking at the report, so we can use a utility function here:</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">TidyMultiqc</span><span class="fu">::</span><span class="fu"><a href="../reference/list_plots.html">list_plots</a></span><span class="op">(</span><span class="va">multiqc_data_path</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 20 × 2</span></span></span>
<span><span class="co">#&gt;    id                                      title                                </span></span>
<span><span class="co">#&gt;    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>                                   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>                                </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 1</span> qualimap_coverage_histogram             Qualimap BamQC: Coverage histogram   </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 2</span> qualimap_genome_fraction                Qualimap BamQC: Genome fraction cove…</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 3</span> qualimap_insert_size                    Qualimap BamQC: Insert size histogram</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 4</span> qualimap_gc_content                     Qualimap BamQC: GC content distribut…</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 5</span> snpeff_variant_effects_region           SnpEff: Counts by Genomic Region     </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 6</span> snpeff_variant_effects_impact           SnpEff: Counts by Effects Impact     </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 7</span> snpeff_effects                          SnpEff: Counts by Effect Types       </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 8</span> snpeff_variant_effects_class            SnpEff: Counts by Functional Class   </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 9</span> snpeff_qualities                        SnpEff: Qualities                    </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">10</span> gatk_varianteval_variant_plot           GATK VariantEval: Variant Counts     </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">11</span> picard_deduplication                    Picard: Deduplication Stats          </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">12</span> fastqc_sequence_counts_plot             FastQC: Sequence Counts              </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">13</span> fastqc_per_base_sequence_quality_plot   FastQC: Mean Quality Scores          </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">14</span> fastqc_per_sequence_quality_scores_plot FastQC: Per Sequence Quality Scores  </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">15</span> fastqc_per_sequence_gc_content_plot     FastQC: Per Sequence GC Content      </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">16</span> fastqc_per_base_n_content_plot          FastQC: Per Base N Content           </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">17</span> fastqc_sequence_duplication_levels_plot FastQC: Sequence Duplication Levels  </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">18</span> fastqc_overrepresented_sequencesi_plot  FastQC: Overrepresented sequences    </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">19</span> fastqc_adapter_content_plot             FastQC: Adapter Content              </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">20</span> fastqc-status-check-heatmap             FastQC: Status Checks</span></span></code></pre>
<p>Now, we know we want the “Per Sequence Quality Scores” plot, and by looking at the data frame above we can tell that the corresponding ID is <code>fastqc_per_sequence_quality_scores_plot</code>.</p>
</div>
<div class="section level3">
<h3 id="loading-plots">Loading Plots<a class="anchor" aria-label="anchor" href="#loading-plots"></a>
</h3>
<p>Now that we have the plot ID, we can load the plot data. First, we need to tell <code>TidyMultiqc</code> to load include some plots by using <code>sections = "plot"</code> (you can load other sections at the same time, as explained above). Also, we need to pass the plot ID from the previous step into the <code>plots</code> argument:</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">df</span> <span class="op">=</span> <span class="fu">TidyMultiqc</span><span class="fu">::</span><span class="fu"><a href="../reference/load_multiqc.html">load_multiqc</a></span><span class="op">(</span></span>
<span>  <span class="va">multiqc_data_path</span>, </span>
<span>  sections <span class="op">=</span> <span class="st">'plot'</span>,</span>
<span>  plots <span class="op">=</span> <span class="st">"fastqc_per_sequence_quality_scores_plot"</span></span>
<span><span class="op">)</span></span>
<span><span class="va">df</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 6 × 2</span></span></span>
<span><span class="co">#&gt;   metadata.sample_id plot.fastqc_per_sequence_quality_scores_plot</span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>              <span style="color: #949494; font-style: italic;">&lt;list&gt;</span>                                      </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> P4107_1001         <span style="color: #949494;">&lt;tibble [35 × 2]&gt;</span>                           </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span> P4107_1002         <span style="color: #949494;">&lt;tibble [38 × 2]&gt;</span>                           </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">3</span> P4107_1003         <span style="color: #949494;">&lt;tibble [31 × 2]&gt;</span>                           </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">4</span> P4107_1004         <span style="color: #949494;">&lt;tibble [32 × 2]&gt;</span>                           </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">5</span> P4107_1005         <span style="color: #949494;">&lt;tibble [40 × 2]&gt;</span>                           </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">6</span> P4107_1006         <span style="color: #949494;">&lt;tibble [40 × 2]&gt;</span></span></span></code></pre></div>
<p>We now have the plot data, but it’s not in a very usable form! This is because each sample has an entire data frame of plot data. At this point if you’re comfortable using <code>dplyr</code> and <code>tidyr</code> to deal with nested data frames, you probably know what to do. Otherwise, read on.</p>
</div>
<div class="section level3">
<h3 id="converting-plot-data">Converting Plot Data<a class="anchor" aria-label="anchor" href="#converting-plot-data"></a>
</h3>
<p>Recall that we are after the median quality score of each sample. First, we should look at the plot data for a single sample to know what we’re dealing with:</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">df</span><span class="op">$</span><span class="va">plot.fastqc_per_sequence_quality_scores_plot</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 35 × 2</span></span></span>
<span><span class="co">#&gt;        x      y</span></span>
<span><span class="co">#&gt;    <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 1</span>     7      2</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 2</span>     8     13</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 3</span>     9    121</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 4</span>    10    702</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 5</span>    11   <span style="text-decoration: underline;">3</span>962</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 6</span>    12  <span style="text-decoration: underline;">11</span>946</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 7</span>    13  <span style="text-decoration: underline;">33</span>566</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 8</span>    14  <span style="text-decoration: underline;">77</span>653</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 9</span>    15 <span style="text-decoration: underline;">157</span>422</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">10</span>    16 <span style="text-decoration: underline;">294</span>209</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># … with 25 more rows</span></span></span></code></pre></div>
<p>So each data frame is a set of x, y pairs. As it’s a histogram plot, we know that the <code>x</code> value is the quality score, and <code>y</code> is the number of times that score has been counted.</p>
</div>
<div class="section level3">
<h3 id="unnesting">Unnesting<a class="anchor" aria-label="anchor" href="#unnesting"></a>
</h3>
<p>One possible way to process this nested data is to use <code>tidyr</code>:</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">df</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu">tidyr</span><span class="fu">::</span><span class="fu"><a href="https://tidyr.tidyverse.org/reference/nest.html" class="external-link">unnest</a></span><span class="op">(</span>cols <span class="op">=</span> <span class="va">plot.fastqc_per_sequence_quality_scores_plot</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 216 × 3</span></span></span>
<span><span class="co">#&gt;    metadata.sample_id     x      y</span></span>
<span><span class="co">#&gt;    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>              <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 1</span> P4107_1001             7      2</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 2</span> P4107_1001             8     13</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 3</span> P4107_1001             9    121</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 4</span> P4107_1001            10    702</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 5</span> P4107_1001            11   <span style="text-decoration: underline;">3</span>962</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 6</span> P4107_1001            12  <span style="text-decoration: underline;">11</span>946</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 7</span> P4107_1001            13  <span style="text-decoration: underline;">33</span>566</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 8</span> P4107_1001            14  <span style="text-decoration: underline;">77</span>653</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 9</span> P4107_1001            15 <span style="text-decoration: underline;">157</span>422</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">10</span> P4107_1001            16 <span style="text-decoration: underline;">294</span>209</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># … with 206 more rows</span></span></span></code></pre></div>
<p>As you can see, if we unnest in this way, we now have multiple rows for the same sample, which is a bit confusing (and not Tidy). However, if we use <code>group_by</code> and then <code>summarise</code>, this can be a useful way to calculate summary statistics. For example, if we want the total number of reads, we could do the following:</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">df</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu">tidyr</span><span class="fu">::</span><span class="fu"><a href="https://tidyr.tidyverse.org/reference/nest.html" class="external-link">unnest</a></span><span class="op">(</span>cols <span class="op">=</span> <span class="va">plot.fastqc_per_sequence_quality_scores_plot</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by</a></span><span class="op">(</span><span class="va">metadata.sample_id</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/summarise.html" class="external-link">summarise</a></span><span class="op">(</span>total_reads <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="va">y</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 6 × 2</span></span></span>
<span><span class="co">#&gt;   metadata.sample_id total_reads</span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>                    <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> P4107_1001           383<span style="text-decoration: underline;">592</span>756</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span> P4107_1002           430<span style="text-decoration: underline;">185</span>524</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">3</span> P4107_1003           431<span style="text-decoration: underline;">379</span>000</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">4</span> P4107_1004           498<span style="text-decoration: underline;">165</span>399</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">5</span> P4107_1005           484<span style="text-decoration: underline;">233</span>426</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">6</span> P4107_1006           453<span style="text-decoration: underline;">238</span>610</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="mapping-with-purrr">Mapping with Purrr<a class="anchor" aria-label="anchor" href="#mapping-with-purrr"></a>
</h3>
<p>Although unnesting worked well in this example, it can get a bit awkward for more complex operations. In these cases we can use <code>purrr</code>. Refer to the next section for an example that compares both approaches. The below example sums the number of reads for each sample, as we have done above, but this time it uses <code><a href="https://purrr.tidyverse.org/reference/map.html" class="external-link">purrr::map_dbl</a></code> to map over the list of data frames:</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">df</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span></span>
<span>    total_reads <span class="op">=</span> <span class="fu">purrr</span><span class="fu">::</span><span class="fu"><a href="https://purrr.tidyverse.org/reference/map.html" class="external-link">map_dbl</a></span><span class="op">(</span><span class="va">plot.fastqc_per_sequence_quality_scores_plot</span>, <span class="op">~</span><span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="va">.</span><span class="op">$</span><span class="va">y</span><span class="op">)</span><span class="op">)</span>,</span>
<span>    plot.fastqc_per_sequence_quality_scores_plot <span class="op">=</span> <span class="cn">NULL</span></span>
<span>  <span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 6 × 2</span></span></span>
<span><span class="co">#&gt;   metadata.sample_id total_reads</span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>                    <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> P4107_1001           383<span style="text-decoration: underline;">592</span>756</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span> P4107_1002           430<span style="text-decoration: underline;">185</span>524</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">3</span> P4107_1003           431<span style="text-decoration: underline;">379</span>000</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">4</span> P4107_1004           498<span style="text-decoration: underline;">165</span>399</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">5</span> P4107_1005           484<span style="text-decoration: underline;">233</span>426</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">6</span> P4107_1006           453<span style="text-decoration: underline;">238</span>610</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="properly-handling-histogram-data">Properly Handling Histogram Data<a class="anchor" aria-label="anchor" href="#properly-handling-histogram-data"></a>
</h3>
<p>Of course, we actually want to find the median here, which is a bit harder. Luckily there exists a package called <code>HistDat</code> for generating summary statistics from histogram-type data. You can check out the package’s manual and vignettes <a href="https://cran.r-project.org/package=HistDat" class="external-link">here</a>, but in brief, we want to convert each of these plot data frames into a <code>HistDat</code> object, which we can do using the same strategies as before. Then, using <code>HistDat</code>, we can calculate our summary statistics in one of the two ways mentioned above.</p>
<p>Using the <code>tidyr</code> approach, we can unnest the plot data, group it, create a <code>HistDat</code> object for each group, and then produce new columns using the new <code>hist</code> column:</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">df</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu">tidyr</span><span class="fu">::</span><span class="fu"><a href="https://tidyr.tidyverse.org/reference/nest.html" class="external-link">unnest</a></span><span class="op">(</span>cols <span class="op">=</span> <span class="va">plot.fastqc_per_sequence_quality_scores_plot</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by</a></span><span class="op">(</span><span class="va">metadata.sample_id</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>hist <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="fu">HistDat</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/HistDat/man/HistDat.html" class="external-link">HistDat</a></span><span class="op">(</span>vals <span class="op">=</span> <span class="va">x</span>, counts <span class="op">=</span> <span class="va">y</span><span class="op">)</span><span class="op">)</span>, .keep <span class="op">=</span> <span class="st">"unused"</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span></span>
<span>    mean_coverage <span class="op">=</span> <span class="va">hist</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/nth.html" class="external-link">first</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/mean.html" class="external-link">mean</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>    median_coverage <span class="op">=</span> <span class="va">hist</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/nth.html" class="external-link">first</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/r/stats/median.html" class="external-link">median</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>    max_coverage <span class="op">=</span> <span class="va">hist</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/nth.html" class="external-link">first</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">max</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>    hist<span class="op">=</span> <span class="cn">NULL</span></span>
<span>  <span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/slice.html" class="external-link">slice</a></span><span class="op">(</span><span class="fl">1</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 6 × 4</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># Groups:   metadata.sample_id [6]</span></span></span>
<span><span class="co">#&gt;   metadata.sample_id mean_coverage median_coverage max_coverage</span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>                      <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>           <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>        <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> P4107_1001                  37.4              40           41</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span> P4107_1002                  37.9              40           41</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">3</span> P4107_1003                  39.4              41           41</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">4</span> P4107_1004                  40.0              41           41</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">5</span> P4107_1005                  38.3              40           41</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">6</span> P4107_1006                  37.6              40           41</span></span></code></pre></div>
<p>Alternatively, using the <code>purrr</code> method, we can just map each plot data frame into a row of summary statistics. Much neater!</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">df</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span></span>
<span>    <span class="fu">purrr</span><span class="fu">::</span><span class="fu"><a href="https://purrr.tidyverse.org/reference/map.html" class="external-link">map_dfr</a></span><span class="op">(</span><span class="va">plot.fastqc_per_sequence_quality_scores_plot</span>, <span class="kw">function</span><span class="op">(</span><span class="va">plot_df</span><span class="op">)</span><span class="op">{</span></span>
<span>      <span class="va">hist</span> <span class="op">=</span> <span class="fu">HistDat</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/HistDat/man/HistDat.html" class="external-link">HistDat</a></span><span class="op">(</span>vals<span class="op">=</span><span class="va">plot_df</span><span class="op">$</span><span class="va">x</span>, counts <span class="op">=</span> <span class="va">plot_df</span><span class="op">$</span><span class="va">y</span><span class="op">)</span></span>
<span>      <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>        mean_coverage <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/mean.html" class="external-link">mean</a></span><span class="op">(</span><span class="va">hist</span><span class="op">)</span>,</span>
<span>        median_coverage <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/median.html" class="external-link">median</a></span><span class="op">(</span><span class="va">hist</span><span class="op">)</span>,</span>
<span>        max_coverage <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">max</a></span><span class="op">(</span><span class="va">hist</span><span class="op">)</span></span>
<span>      <span class="op">)</span></span>
<span>    <span class="op">}</span><span class="op">)</span>,</span>
<span>    plot.fastqc_per_sequence_quality_scores_plot <span class="op">=</span> <span class="cn">NULL</span></span>
<span>  <span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 6 × 4</span></span></span>
<span><span class="co">#&gt;   metadata.sample_id mean_coverage median_coverage max_coverage</span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>                      <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>           <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>        <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> P4107_1001                  37.4              40           41</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span> P4107_1002                  37.9              40           41</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">3</span> P4107_1003                  39.4              41           41</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">4</span> P4107_1004                  40.0              41           41</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">5</span> P4107_1005                  38.3              40           41</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">6</span> P4107_1006                  37.6              40           41</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="custom-plot-parsers">Custom Plot Parsers<a class="anchor" aria-label="anchor" href="#custom-plot-parsers"></a>
</h3>
<p>So far, we have used <code>TidyMultiqc</code>’s built-in parsers. Each parser is a function that converts the plot JSON into a list of data frames. However, it is possible that we might encounter a new type of plot that is not yet implemented in <code>TidyMultiqc</code>. If this happens, the first thing you should do is <a href="https://github.com/multimeric/TidyMultiqc/issues" class="external-link">file an issue against this package</a>.</p>
<p>Then, if you’re daring, you can try to implement a parser. For your reference, you can refer to the existing parsers in the source code. Then, you can pass in your custom parser using the <code>plot_parsers</code> argument to <code>load_multiqc</code>:</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">TidyMultiqc</span><span class="fu">::</span><span class="fu"><a href="../reference/load_multiqc.html">load_multiqc</a></span><span class="op">(</span></span>
<span>  <span class="va">multiqc_data_path</span>, </span>
<span>  sections <span class="op">=</span> <span class="st">'plot'</span>,</span>
<span>  plots <span class="op">=</span> <span class="st">"fastqc_per_sequence_quality_scores_plot"</span>,</span>
<span>  plot_parsers <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>    <span class="co"># This fake parser function takes a plot and just returns the iris dataset</span></span>
<span>    xy_line <span class="op">=</span> <span class="kw">function</span><span class="op">(</span><span class="va">plot_data</span>, <span class="va">name</span><span class="op">)</span><span class="op">{</span></span>
<span>      <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>        sample_1 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>          plot_name <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">iris</span><span class="op">)</span></span>
<span>        <span class="op">)</span></span>
<span>      <span class="op">)</span></span>
<span>    <span class="op">}</span></span>
<span>  <span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 1 × 2</span></span></span>
<span><span class="co">#&gt;   metadata.sample_id plot_name     </span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>              <span style="color: #949494; font-style: italic;">&lt;list&gt;</span>        </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> sample_1           <span style="color: #949494;">&lt;df [150 × 5]&gt;</span></span></span></code></pre></div>
<p>Finally, if your parser works, please submit a pull request to <code>TidyMultiqc</code> to share this with everyone!</p>
<p>Using the strategies and patterns explained in this section, you should be in good stead to handle whatever plots MultiQC throws at you.</p>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Michael Milton.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
